xen/arch/arm/coproc/coproc.c

```diff
@@ -42,23 +43,45 @@ s_time_t vcoproc_context_switch(struct vcoproc_instance *curr,
                                 struct vcoproc_instance *next)
 {
     struct coproc_device *coproc;
-    s_time_t wait_time;
     int ret;
 
     if ( unlikely(curr == next) )
         return 0;
 
     coproc = next ? next->coproc : curr->coproc;
 
-    wait_time = coproc->ops->ctx_switch_from(curr);
-    if ( wait_time )
-        return wait_time;
+    if ( likely(curr) )
+    {
+        s_time_t wait_time;
 
-    /* TODO What to do if we failed to switch to "next"? */
-    ret = coproc->ops->ctx_switch_to(next);
-    if ( ret )
-        panic("Failed to switch context to vcoproc \"%s\" (%d)\n",
-              dev_path(coproc->dev), ret);
+        ASSERT(curr->state == VCOPROC_RUNNING ||
+               curr->state == VCOPROC_ASKED_TO_SLEEP);
+
+        wait_time = coproc->ops->ctx_switch_from(curr);
+
+        if ( wait_time == 0 )
+        {
+            if (curr->state == VCOPROC_RUNNING)
+                curr->state = VCOPROC_WAITING;
+            else
+                curr->state = VCOPROC_SLEEPING;
+        }
+        if ( wait_time )
```

Comment:
won't fix: original logic preserved. such changes should be in a separate change-set. I'll open an issue for that